### ISSUE TITLE
libnd4j: multi_head_dot_product_attention and ..._bp memory use

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#8345]
Running simple tests here and tracking GPU memory use:

https://gist.github.com/AlexDBlack/e663474654e8bf8c43761fe5ad572198

![attention_mem_use](https://user-images.githubusercontent.com/2360237/68097877-9d495f80-ff0d-11e9-8bab-04750dc7fe11.png)

Interpreting this graph:
* The flat section (x=11 to 21) is the java allocated input and output arrays + 10 second sleep
* All memory changes (allocations and deallocations) after this are done within the op itself - at C++ level only

Memory use:
* Forward: before=2346mb, peak=3212mb -> C++ working memory: 866 MB
* Backward: before=2877mb, peak=7663mb -> C++ working memory: 4786 MB

Obviously, there are 2 problems with this:
(a)The memory use requirements are very large for these ops, and we should optimize this
(b) Large allocation and deallocations generally are costly; some sort of workspace system may be needed for the C++ op working memory to improve performance

**Comment by AlexDBlack** [comment]
So, multi-head attention forward pass. There's definitely a few things we can do here:
* Special case for self attention: if query == key, or key == value, then just do one
    * Easy to check: just same pointer, shape and strides
    * Saves 160MB here
* Remove the (usually) unnecessary assign here: https://github.com/eclipse/deeplearning4j/blob/master/libnd4j/include/ops/declarable/generic/nn/multi_head_dot_product_attention.cpp#L116
    * Saves 20MB in this case
* Deallocate the 3 projected* arrays immediately after execution of this op: https://github.com/eclipse/deeplearning4j/blob/master/libnd4j/include/ops/declarable/generic/nn/multi_head_dot_product_attention.cpp#L103
    * May not help reduce peak memory requirements though, as we need them for the dot product attention op
* Switch to a "loop over heads" design (see below)

Working memory requirements for current multi-head attention op, forward case:
`{mb, nHeads, projSize, seqLen}` x3, for the input projection matrices - 80MB x3 for float32, for the above case
The (pre output mmul) projection result array (attnResults) is 80MB here also.
Total working memory by my count is 340MB plus whatever is used in the (single headed) attention op.

Loop over heads design: quite simple.
Figure 2 (page 4) here: http://papers.nips.cc/paper/7181-attention-is-all-you-need.pdf

Note that the the attention heads are running in parallel except for the "concat -> linear" section.
The linear is with the output projection matrix, shape `[projSize*headSize,  nOut]`
and the output of "Concat" has shape `[minibatch*seqLength, projSize*nHeads]`
The output of the linear thus has shape `[minibatch*seqLength, nOut]` (2d reshaped time series)
i.e., `[minibatch*seqLength, projSize*nHeads] x [projSize*headSize,  nOut]`
Effectively, the concatenation is along dimension 1 here.

If we split it up, instead of concat + linear we simply do:
```
out = zeros(minibatch*seqLength, nOut)
for( int i=0; i<nHeads; i++ ){
    singleHeadAttention = ...
    out += singleHeadAttention x WOut.get(interval(i * projSize, (i+1)*projSize), all))
}
```

This approach should reduce our working memory use by a factor of nHeads - so a factor of 4, in this case, so something like 600mb reduction.
Only real downside I can see is that we have a larger number of smaller operations (mmul and others), hence maybe slightly worse hardware utilization/efficiency.

Single head attention (dot product attention) looks OK to me, I don't think any optimizations there is needed.

I'll look at attention backprop soon. But based on the structure, we should be able to do a similar thing as here.

**Comment by agibsonccc** [comment]
Closing — ND4J/libnd4j core infrastructure has been significantly rewritten (new memory management, op execution engine, backend architecture). If this issue is still relevant with the upcoming release, please open a new issue.
